File: 백준/Silver/9012. 괄호/괄호.js
// /dev/stdin
const fs = require("fs");
let input = fs
  .readFileSync("/dev/stdin")
  .toString()
  .split("\n")
  .map((x) => x.replace("\r", ""));

const N = Number(input[0]);
let answer = "";

for (let i = 1; i <= N; i++) {
  const word = input[i];
  const stack = [];
  let isVPS = true;
  for (const v of word) {
    if (v === "(") {
      stack.push("(");
      continue;
    }

    if (v === ")") {
      if (stack.length === 0) {
        isVPS = false;
        break;
      } else {
        stack.pop();
      }

      continue;
    }
  }
  //console.log(stack);
  if (isVPS && stack.length === 0) {
    answer += "YES \n";
  } else {
    answer += "NO \n";
  }
}

console.log(answer);

File: 백준/Silver/9012. 괄호/괄호.py
import sys


T = int(sys.stdin.readline().strip())

for _ in range(T):
  word = list(sys.stdin.readline().strip())
  stack = []
  answer = 'YES'


  for s in word:
    if s == '(':
      stack.append('(')
    elif s == ')':
      if stack and stack[-1] == '(':
        stack.pop()
      elif len(stack) == 0:
        answer = 'NO'
        break
      
  if len(stack) > 0:
    answer = 'NO'
  print(answer)

      


  # break
